### ISSUE TITLE
ServiceMonitorSelector labels are not honored correctly

### ISSUE THREAD

**Issue by missourian55** [open-telemetry/opentelemetry-operator#1595]
I am facing this strange issue and here are the steps to reproduce the bug.

1. Deploy *two distinct collectors*, one with ```serviceMonitorSelector``` configuration and the other *without it*

2. Deploy a ```ServiceMonitor``` configuration without the  label attribute specified in the `serviceMonitorSelector`.

3. Both the collectors  logs shows that  *no targets have been registered*.

4. Add the label specified in the `serviceMonitorSelector` to the `ServiceMonitor` and redeploy.

5. Verify in the logs of both collectors that targets are registered on both the collectors, the one with serviceMonitorSelector and the collector without it. 

**One with ServiceMonitorSelector**

```yaml
---
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: adot-1
  namespace: default
spec:
  mode: statefulset
  serviceAccount: collector-sa
  targetAllocator:
    enabled: true
    serviceAccount: allocator-sa
    filterStrategy: relabel-config
    prometheusCR:
      enabled: true
      serviceMonitorSelector:
        env: production
  env:
    - name: KUBERNETES_NAMESPACE
      valueFrom:
        fieldRef:
          fieldPath: metadata.namespace
  config: |
    receivers:
      prometheus:
        target_allocator:
          endpoint: "http://adot-1-targetallocator"
          interval: 30s
          collector_id: ${POD_NAME}
        config:
          scrape_configs:
          - job_name: 'otel-collector'
            scrape_interval: 1000s
            static_configs:
            - targets: [ '0.0.0.0:8888' ]
    
    processors:
      batch/metrics:
        timeout: 60s  

    exporters:
      logging:
        loglevel: debug

    service:
      pipelines:
        metrics:
          receivers: [prometheus]
          processors: [batch/metrics]
          exporters: [logging]
---
```

**One without ServiceMonitorSelector**

```yaml
---
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: adot-2
  namespace: default
spec:
  mode: statefulset
  serviceAccount: collector-sa
  targetAllocator:
    enabled: true
    serviceAccount: allocator-sa
    filterStrategy: relabel-config
    prometheusCR:
      enabled: true
  env:
    - name: KUBERNETES_NAMESPACE
      valueFrom:
        fieldRef:
          fieldPath: metadata.namespace
  config: |
    receivers:
      prometheus:
        target_allocator:
          endpoint: "http://adot-2-targetallocator"
          interval: 30s
          collector_id: ${POD_NAME}
        config:
          scrape_configs:
          - job_name: 'otel-collector'
            scrape_interval: 1000s
            static_configs:
            - targets: [ '0.0.0.0:8888' ]
    
    processors:
      batch/metrics:
        timeout: 60s  

    exporters:
      logging:
        loglevel: debug
    
    service:
      pipelines:
        metrics:
          receivers: [prometheus]
          processors: [batch/metrics]
          exporters: [logging]
---
```

**Sample ServiceMonitor for testing**

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: demo-app-monitor
  namespace: test
  labels:
    env: production
spec:
  selector:
    matchLabels:
      app.kubernetes.io/name: demo-app
      app.kubernetes.io/version: 1.0
  namespaceSelector:
    matchNames:
      - test
  endpoints:
    - port: http
      path: /metrics
      interval: 20s
      scrapeTimeout: 10s
```

**Comment by pavolloffay** [comment]
cc) @jaronoff97 @kristinapathak

**Comment by kristinapathak** [comment]
Step 5 (both collectors registering the `ServiceMonitor`) matches my understanding of the current functionality.  I'm not sure why the collector without the `serviceMonitorSelector` set did not find and register the `ServiceMonitor` in step 3.  I will try to reproduce that for myself tomorrow and investigate why that's happening.

The reason why the collector without any `serviceMonitorSelector` set is finding the `ServiceMonitor` is because `prometheusCR` is still enabled:

1. We [set up](https://github.com/open-telemetry/opentelemetry-operator/blob/d90c913756f4e7cc6c9ab2f2e0dceae80907d76a/cmd/otel-allocator/main.go#L106) a CR Watcher in the target allocator because it is enabled.
2. As part of `NewPrometheusCRWatcher()`, we [create selectors ](https://github.com/open-telemetry/opentelemetry-operator/blob/d90c913756f4e7cc6c9ab2f2e0dceae80907d76a/cmd/otel-allocator/watcher/promOperator.go#L64) from the service monitor selector config.
3. I'm pretty sure that unmarshaling an unset service monitor selector map leads to it being nil (verified [here](https://go.dev/play/p/jZGZd23LxO1)), which means we return the result of `labels.NewSelector()` [here](https://github.com/open-telemetry/opentelemetry-operator/blob/d90c913756f4e7cc6c9ab2f2e0dceae80907d76a/cmd/otel-allocator/watcher/promOperator.go#L93).  From the subsequent kubernetes [labels pkg](https://github.com/kubernetes/apimachinery/blob/53ecdf01b997ca93c7db7615dfe7b27ad8391983/pkg/labels/selector.go#LL107C18-L107C18), this is a nil `internalSelector`.
4. Regardless of the value of the selector, we set it [here](https://github.com/open-telemetry/opentelemetry-operator/blob/d90c913756f4e7cc6c9ab2f2e0dceae80907d76a/cmd/otel-allocator/watcher/promOperator.go#L74), and then use it in the `ListAll()` function [later](https://github.com/open-telemetry/opentelemetry-operator/blob/d90c913756f4e7cc6c9ab2f2e0dceae80907d76a/cmd/otel-allocator/watcher/promOperator.go#L139).
5. The `ListAll()` function [leads](https://github.com/prometheus-operator/prometheus-operator/blob/97852301dbafafb00e08fed404d8467aec673031/pkg/informers/informers.go#L110) us [here](https://github.com/kubernetes/client-go/blob/8cbca742aebe24b24f7f4e32fd999942fa9133e8/tools/cache/listers.go#L35), where we check if the selector is empty.  [It is](https://github.com/kubernetes/apimachinery/blob/53ecdf01b997ca93c7db7615dfe7b27ad8391983/pkg/labels/selector.go#L309).  In this [case](https://github.com/kubernetes/client-go/blob/8cbca742aebe24b24f7f4e32fd999942fa9133e8/tools/cache/listers.go#L37-L42), everything is selected.

If the service monitor selector config is somehow an empty map instead of a nil map, we would reach the kubernetes [labels pkg function](https://pkg.go.dev/k8s.io/apimachinery/pkg/labels#SelectorFromSet) `SelectorFromSet()` which returns a selector that selects everything.  This will effectively give us the same result as a nil config map.

If we wanted to select [nothing](https://pkg.go.dev/k8s.io/apimachinery/pkg/labels#Nothing) instead of everything for both a nil and empty map, we can change the `getSelector()` [func](https://github.com/open-telemetry/opentelemetry-operator/blob/d90c913756f4e7cc6c9ab2f2e0dceae80907d76a/cmd/otel-allocator/watcher/promOperator.go#L90) to include:
```go
if len(s) == 0 {
    return labels.Nothing()
}
```

If you follow along in the code, you can see the same functionality existing for the pod monitor selector. I can't find anything in our documentation indicating what happens when the `serviceMonitorSelector` and/or `podMonitorSelector` are empty.  Given the phrasing [here](https://github.com/open-telemetry/opentelemetry-operator/blob/d90c913756f4e7cc6c9ab2f2e0dceae80907d76a/apis/v1alpha1/opentelemetrycollector_types.go#LL219-L222) of "ServiceMonitors to be selected for target discovery." and "The requirements are ANDed.", this may imply that we should be selecting nothing when no requirements are given? Or it may mean I need to take a small break. 🙂 

---
**TLDR**: when prometheusCR is enabled, `serviceMonitorSelector` being empty means all `ServiceMonitor`s are registered.  The same functionality is there for `podMonitorSelector`.  If we wanted to, it would be easy to flip this so unset monitor selectors lead to nothing being registered. Either way, we should update the documentation to be more clear.

**Comment by jaronoff97** [comment]
+1 to what kristina said, the issue with flipping the behavior (and why this wasn't done when we implemented the feature) was because it would break any existing installation without upgrades. This would be a good thing to feature flag most likely :)

**Comment by kristinapathak** [comment]
I wasn't able to reproduce step 3 - when the collector with no service monitor selectors provided didn't find any service monitors even though one existed.  I still think we should update docs to specify the behavior when `serviceMonitorSelector` or `podMonitorSelector` are unset.

**Comment by jaronoff97** [comment]
This will be completed by #2369 and as part of our v2 milestone.

**Comment by jaronoff97** [comment]
closing this issue in favor of https://github.com/open-telemetry/opentelemetry-operator/issues/1907. we've worked on a permanent fix for this which will be available in the v1beta1 version of our CRDs.
